p2p/ipld/sample.go

```diff
@@ -0,0 +1,99 @@
+package ipld
+
+import (
+	"github.com/ipfs/go-cid"
+	"github.com/lazyledger/nmt/namespace"
+
+	"github.com/lazyledger/lazyledger-core/libs/rand"
+	"github.com/lazyledger/lazyledger-core/p2p/ipld/plugin/nodes"
+	"github.com/lazyledger/lazyledger-core/types"
+)
+
+// Sample is a point in 2D space over square.
+type Sample struct {
+	Row, Col uint32
+}
+
+// SampleSquare randomly picks *num* unique points from arbitrary *width* square
+// and returns them as samples.
+func SampleSquare(squareWidth uint32, num int) []*Sample {
+	ss := newSquareSampler(squareWidth, num)
+	ss.sample(num)
+	return ss.sampled()
+}
+
+// Leaf returns leaf info needed for retrieval using data provided with DAHeader.
+func (s *Sample) Leaf(dah *types.DataAvailabilityHeader) (cid.Cid, uint32, error) {
+	var (
+		leaf uint32
+		root namespace.IntervalDigest
+	)
+
+	// spread leaves retrieval from both Row and Column roots
+	if rand.Bool() {
+		root = dah.ColumnRoots[s.Col]
+		leaf = s.Row
+	} else {
+		root = dah.RowsRoots[s.Row]
+		leaf = s.Col
+	}
```

Comment:
@adlerjohn, what do you think about this? Will that improve confidence in availability validity?

Reply:
In short, as we can get the same samples from rows and columns, downloading them only from e.g columns may introduce a shortage in row non-leaf nodes distribution over the network, as all the light clients would depend only on columns.

Reply:
Or we can always retrieve a sample from both places

Reply:
ref: https://github.com/lazyledger/lazyledger-specs/issues/110#issuecomment-818786351

Reply:
@liamsi, Do we want to include a decision for an issue above within the PR?

Reply:
IMO, it's OK to do this in a separate followup PR. Also, I'm also not sure if @adlerjohn will work on it this week.

Ideally, after the branch gets updated and merge conflicts are resolved, we can merge this PR :-)